### PR TITLE
Integrate docker-slim for waaaaaaaay smaller run images

### PR DESCRIPTION
**Pull request by lreiher**
*docker-ros* now integrates the [*slim*](https://github.com/slimtoolkit/slim) toolkit for minifying container images. *slim* is enabled by default and will, in addition to the `run` deployment image, produce an additional `:latest-slim`-tagged minified image. Note that the slimmed deployment image is stripped of every single thing not needed for executing the default launch command. The slimming process can be controlled via the `SLIM_BUILD_ARGS` configuration variable.

### CODE REVIEW

File: .gitlab-ci/docker-ros.yml [hunk 1]
```diff
@@ -71,19 +80,20 @@ stages:
   - Build dev Images
   - Build run Images
   - Test ROS Industrial CI
+  - Slim Images
   - Push Multi-Arch Images
 
 
 default:
-  image: docker:20.10.22-git
+  image: docker:20.10.22-git # TODO: upgrade?
```
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:88]: upgrade?

File: .gitlab-ci/docker-ros.yml [hunk 2]
```diff
@@ -231,6 +238,55 @@ Test run-arm64:
...
+  stage: Slim Images
+  before_script:
+    - apk add --update --upgrade curl tar
+    - curl -L -o ds.tar.gz ${SLIM_DOWNLOAD_URL}
+    - tar -xvf ds.tar.gz
+    - cd dist_linux*
+    - docker login -u ${REGISTRY_USER} -p ${REGISTRY_PASSWORD} ${REGISTRY}
+    - docker pull ${FAT_IMAGE}
+  script:
+    - ./slim build --sensor-ipc-mode proxy --target ${FAT_IMAGE} --tag ${SLIM_IMAGE} --continue-after=10 --show-clogs --http-probe=false
```
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:251]: Add config variable to pass any argument to slim.

File: .gitlab-ci/docker-ros.yml
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:1]: Implement for GitHub Action

File: .gitlab-ci/docker-ros.yml [hunk 3]
```diff
@@ -17,6 +17,8 @@ variables:
   IMAGE_TAG:                  latest                                  # Image tag of run image
   DEV_IMAGE_NAME:             ${IMAGE_NAME}                           # Image name of dev image
   DEV_IMAGE_TAG:              ${IMAGE_TAG}-dev                        # Image tag of dev image
+  SLIM_IMAGE_NAME:            ${IMAGE_NAME}                           # Image name of slim run image
+  SLIM_IMAGE_TAG:             ${IMAGE_TAG}-slim                       # Image tag of slim run image
```
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:21]: Document new variables and feature in general in README.

File: .gitlab-ci/docker-ros.yml [hunk 4]
```diff
@@ -231,6 +238,55 @@ Test run-arm64:
     _PLATFORM: arm64
 
 
+.slim:
+  stage: Slim Images
```
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:242]: Add `slim` as a `TARGET`?

File: .gitlab-ci/docker-ros.yml [hunk 5]
```diff
@@ -47,19 +49,26 @@ variables:
...
   _IMAGE_RUN_CI:          ${_RUN_IMAGE}_${CI_COMMIT_REF_SLUG}_ci
+  _IMAGE_SLIM_CI:         ${_SLIM_IMAGE}_${CI_COMMIT_REF_SLUG}_ci
   _IMAGE_DEV_CI_AMD64:    ${_IMAGE_DEV_CI}-amd64
   _IMAGE_DEV_CI_ARM64:    ${_IMAGE_DEV_CI}-arm64
   _IMAGE_RUN_CI_AMD64:    ${_IMAGE_RUN_CI}-amd64
   _IMAGE_RUN_CI_ARM64:    ${_IMAGE_RUN_CI}-arm64
-  _IMAGE_DEV_LATEST:      ${DEV_IMAGE_NAME}:latest-dev
+  _IMAGE_SLIM_CI_AMD64:   ${_IMAGE_SLIM_CI}-amd64
+  _IMAGE_SLIM_CI_ARM64:   ${_IMAGE_SLIM_CI}-arm64
+  _IMAGE_DEV_LATEST:      ${DEV_IMAGE_NAME}:latest-dev # TODO: is this a bug here that DEV_IMAGE_TAG is not used?
```
Comment (jpbusch) [inline .gitlab-ci/docker-ros.yml:63]: yes, seems so

File: .gitlab-ci/docker-ros.yml [hunk 6]
```diff
@@ -47,19 +49,26 @@ variables:
...
   _IMAGE_RUN_LATEST:      ${IMAGE_NAME}:latest
+  _IMAGE_SLIM_LATEST:     ${SLIM_IMAGE_NAME}:latest-slim
   _IMAGE_DEV_TARGET_TAG:  ${_DEV_IMAGE}-${CI_COMMIT_TAG}
   _IMAGE_RUN_TARGET_TAG:  ${_RUN_IMAGE}-${CI_COMMIT_TAG}
-  _IMAGE_DEV_TAG:         ${DEV_IMAGE_NAME}:${CI_COMMIT_TAG}-dev
+  _IMAGE_SLIM_TARGET_TAG: ${_SLIM_IMAGE}-${CI_COMMIT_TAG}
+  _IMAGE_DEV_TAG:         ${DEV_IMAGE_NAME}:${CI_COMMIT_TAG}-dev # TODO: is this a bug here that DEV_IMAGE_TAG is not used?
```
Comment (jpbusch) [inline .gitlab-ci/docker-ros.yml:69]: here this should be correct

File: .gitlab-ci/docker-ros.yml (hunk 1 above)
Comment (jpbusch) [reply .gitlab-ci/docker-ros.yml:88]: I would suggest to test and upgrade this within a separate PR.

File: .gitlab-ci/docker-ros.yml
Comment (jpbusch) [reply .gitlab-ci/docker-ros.yml:1]: Seems there is already an GitHub Action, which we can use: https://github.com/marketplace/actions/docker-slim-github-action

Comment (lreiher) [reply .gitlab-ci/docker-ros.yml:1]: May be difficult then to allow the docker-ros user to pass any slim arguments.

File: .gitlab-ci/docker-ros.yml (hunk 5 above)
Comment (lreiher) [reply .gitlab-ci/docker-ros.yml:63]: Actually not a bug, I think. `_IMAGE_DEV_LATEST` refers to the dev image that is pushed in the case that `push-as-latest` is active, a flag to also push `latest` even though a different tag is normally set.

We offer variables to control run tag and dev tag separately, but not for only controlling the `-dev` suffix. If we did, we could also consider using it here, but we don't. So I think we can keep it as-is.

File: .gitlab-ci/docker-ros.yml (hunk 6 above)
Comment (lreiher) [reply .gitlab-ci/docker-ros.yml:69]: Actually not a bug, I think. `_IMAGE_DEV_TAG` refers to the dev image that is tagged with a Git tag.

We offer variables to control run tag and dev tag separately, but not for only controlling the `-dev` suffix. If we did, we could also consider using it here, but we don't. So I think we can keep it as-is.

File: .gitlab-ci/docker-ros.yml (hunk 4 above)
Comment (lreiher) [reply .gitlab-ci/docker-ros.yml:242]: Would have to adjust logic of TARGET, so we decided to use `ENABLE_SLIM` flag instead.

File: scripts/ci.sh [hunk 7]
```diff
@@ -75,4 +84,23 @@ for PLATFORM in "${PLATFORMS[@]}"; do
...
+
+    # slim image
+    if [[ "${ENABLE_SLIM}" == "true" && "${TARGET}" == "run" ]]; then
+        open_log_group "Slim image (${PLATFORM})"
+        image="${IMAGE}"
+        slim_image="${SLIM_IMAGE}"
+        [[ -n "${_IMAGE_POSTFIX}" ]] && image="${image}${_IMAGE_POSTFIX}"
+        [[ -n "${_IMAGE_POSTFIX}" ]] && slim_image="${slim_image}${_IMAGE_POSTFIX}"
+        [[ "${_ENABLE_IMAGE_PUSH}" != "true" || "${ENABLE_SINGLEARCH_PUSH}" == "true" ]] && image="${image}-${PLATFORM}"
+        # TODO: not yet working; builds image with suffix, pushed without? [[ "${_ENABLE_IMAGE_PUSH}" != "true" || "${ENABLE_SINGLEARCH_PUSH}" == "true" ]] && slim_image="${slim_image}-${PLATFORM}"
```
Comment (lreiher) [inline scripts/ci.sh:98]: TODO

**Pull request merged**